svcb.go

```diff
@@ -0,0 +1,795 @@
+//	o.Value = append(o.Value, e)
+type SVCBLocal struct {
+	KeyCode uint16 // Never 65535 or any assigned keys
+	Data    []byte // All byte sequences are allowed
+	// For the string representation, See draft-ietf-dnsop-svcb-https
+	// (TODO RFC XXXX)
+	// "2.1.1.  Presentation format for SVCBFieldValue key=value pairs"
+	// for a full list of allowed characters. Otherwise escape codes
+	// e.g. \000 for NUL and \127 for DEL are used.
+}
+
+func (s *SVCBLocal) Key() uint16           { return s.KeyCode }
+func (s *SVCBLocal) pack() ([]byte, error) { return s.Data, nil }
+func (s *SVCBLocal) unpack(b []byte) error { s.Data = b; return nil }
+func (s *SVCBLocal) len() uint16           { return uint16(len(s.Data)) }
+
+// String escapes whitespaces too, which is not required when
+// the result would be enclosed in double quotes. TODO Is this doc fine?
+// do i need definition
+func (s *SVCBLocal) String() string {
+	var str strings.Builder
+	str.Grow(4 * len(s.Data))
+	for _, e := range s.Data {
+		if (0x1f < e && e < 0x7f) || e == 0x09 {
+			switch e {
+			case '"':
+				fallthrough
+			case ';':
+				fallthrough
+			// As promised, optionally handle space
+			case ' ':
+				fallthrough
+			// Tab
+			case 0x09:
+				fallthrough
+			case '\\':
+				str.WriteByte('\\')
+				fallthrough
+			default:
+				str.WriteByte(e)
+			}
+		} else {
+			str.WriteString(escapeByte(e))
+		}
+	}
+	return str.String()
+}
+
+func (s *SVCBLocal) parse(b string) error {
+	bytes := make([]byte, 0, len(b))
+	i := 0
+	for i < len(b) {
+		if b[i] == '\\' {
+			if i+1 == len(b) {
+				return errors.New("dns: SVCB private/experimental key" +
+					" escape unterminated")
+			}
+			if isDigit(b[i+1]) {
+				if i+3 < len(b) && isDigit(b[i+2]) && isDigit(b[i+3]) {
+					a, err := strconv.ParseUint(b[i+1:i+4], 10, 8)
+					if err == nil {
+						i += 4
+						bytes = append(bytes, byte(a))
+						continue
+					}
+				}
+				return errors.New("dns: SVCB private/experimental key" +
+					" bad escaped octet")
+			} else {
+				bytes = append(bytes, b[i+1])
+				i += 2
+			}
+		} else {
+			bytes = append(bytes, b[i])
+			i++
+		}
+	}
+	s.Data = bytes
+	return nil
+}
+
+func (s *SVCBLocal) copy() SVCBKeyValue {
+	return &SVCBLocal{s.KeyCode,
+		append(make([]byte, 0, len(s.Data)), s.Data...),
+	}
+}
+
+func (rr *SVCB) String() string {
+	s := rr.Hdr.String() +
+		strconv.Itoa(int(rr.Priority)) + " " +
+		sprintName(rr.Target)
+	for _, element := range rr.Value {
+		s += " " + svcbKeyToString(element.Key()) +
+			"=\"" + element.String() + "\""
+	}
+	return s
+}
+
+// areSVCBPairArraysEqual checks if SVCBKeyValue arrays are equal
+// after sorting their copies. arrA and arrB have equal lengths,
+// otherwise zduplicate.go wouldn't call this function.
+func areSVCBPairArraysEqual(arrA []SVCBKeyValue, arrB []SVCBKeyValue) bool {
+	a := append(make([]SVCBKeyValue, 0, len(arrA)), arrA...)
+	b := append(make([]SVCBKeyValue, 0, len(arrB)), arrB...)
+	sort.Slice(a, func(i, j int) bool {
```

Comment:
These can probably be one line: `sort.Slice(a, func(i, j int) bool { return a[i].Key() < a[j].Key() })`.

Reply:
`isDuplicate` shouldn't modify its inputs.

Reply:
My suggestion here is just to put the existing code onto one line.